File: src/main/java/cn/labzen/logger/logback/LabzenLogbackLoggerContext.java
package cn.labzen.logger.logback;

import ch.qos.logback.classic.LoggerContext;
import cn.labzen.logger.kernel.LabzenLogger;
import cn.labzen.logger.logback.filter.ForcedFilter;
import org.slf4j.ILoggerFactory;
import org.slf4j.Logger;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class LabzenLogbackLoggerContext implements ILoggerFactory {

  private static volatile LabzenLogbackLoggerContext INSTANCE;
  private static final Map<String, Logger> LOGGER_CACHE = new ConcurrentHashMap<>();
  private final LoggerContext principal;

  public LabzenLogbackLoggerContext(LoggerContext principal) {
    principal.addTurboFilter(new ForcedFilter());
    this.principal = principal;
    INSTANCE = this;
  }

  @Override
  public Logger getLogger(String name) {
    return LOGGER_CACHE.computeIfAbsent(name, key -> {
      Logger original = principal.getLogger(key);
      return new LabzenLogger(original);
    });
  }

  public LoggerContext getContext() {
    return principal;
  }

  public static LabzenLogbackLoggerContext instance() {
    return INSTANCE;
  }
}

File: src/main/kotlin/cn/labzen/logger/logback/LabzenLogbackLoggerContext.kt
package cn.labzen.logger.logback

import ch.qos.logback.classic.LoggerContext
import cn.labzen.logger.kernel.LabzenLogger
import cn.labzen.logger.logback.filter.ForcedFilter
import cn.labzen.logger.logback.pattern.conversion.IdentifiableLoggerConverter
import org.slf4j.ILoggerFactory
import org.slf4j.Logger

@Suppress("unused")
class LabzenLogbackLoggerContext(private val principal: LoggerContext) : ILoggerFactory {

  init {
    instance = this

    principal.addTurboFilter(ForcedFilter())
    IdentifiableLoggerConverter.collectLabzenComponentPackages()
  }

  internal fun getContext() = principal

  override fun getLogger(name: String): Logger {
    val original: Logger = principal.getLogger(name)
    return LabzenLogger(original)
  }

  companion object {

    private var instance: LabzenLogbackLoggerContext? = null

    internal fun singleton(): LabzenLogbackLoggerContext {
      return instance!!
    }
  }
}
